File: src/main/java/fi/quanfoxes/parser/Context.java
package fi.quanfoxes.parser;

import java.util.Collection;
import java.util.HashMap;

import fi.quanfoxes.parser.nodes.TypeNode;

public class Context {
    private Context context;

    private HashMap<String, Variable> variables = new HashMap<>();
    private HashMap<String, Functions> functions = new HashMap<>();
    private HashMap<String, Type> types = new HashMap<>();
    private HashMap<String, Label> labels = new HashMap<>();

    /**
     * Updates types, function and variables when new context is linked
     */
    protected void update() {
        for (Variable variable : getVariables()) {
            if (variable.isTypeUnresolved()) {
                Resolvable resolvable = (Resolvable)variable.getType();

                try {
                    TypeNode type = (TypeNode)resolvable.resolve(this);
                    variable.setType(type.getType());

                } catch (Exception e) {}
            }
        }

        for (Type type : getTypes()) {
            type.update();
        }

        for (Functions entry : getFunctions()) {
            entry.update();
        }
    }

    /**
     * Links this context with the given context, allowing access to the information of the given context
     * @param context Context to link with
     */
    public void link(Context context) {
        this.context = context;
        this.update();
    }

    /**
     * Moves all types, functions and variables from the given context to this context, and then destroyes the given context
     * @param context Context to merge with
     */
    public void merge(Context context) {
        types.putAll(context.types);
        functions.putAll(context.functions);
        variables.putAll(context.variables);

        for (Type type : context.types.values()) {
            type.setParent(this);
        }

        for (Functions entry : context.functions.values()) {
            for (Function function : entry.getFunctions()) {
                function.setParent(this);
            }
        }

        for (Variable variable : context.variables.values()) {
            variable.setContext(this);
        }

        update();
    }

    /**
     * Sets the parent of this context
     * @param context New parent context
     */
    public void setParent(Context context) {
        this.context = context;
    }

    /**
     * Returns the parent of this context
     * @return Parent of this context
     */
    public Context getParent() {
        return context;
    }

    /**
     * Declares new type in this context
     * @param type Type to declare
     * @throws Exception When type with same name already exists in this context this function throws an exception
     */
    public void declare(Type type) throws Exception {
        if (isLocalTypeDeclared(type.getName())) {
            throw new Exception(String.format("Type '%s' already exists in this context", type.getName()));
        }

        types.put(type.getName(), type);
    }

    /**
     * Declares new function in this context
     * @param function Function to declare
     */
    public void declare(Function function) throws Exception {
        Functions entry;

        if (isLocalFunctionDeclared(function.getName())) {
            entry = functions.get(function.getName());
        }
        else {
            functions.put(function.getName(), (entry = new Functions()));
        }

        entry.add(function);
    }

    /**
     * Declares new variable in this context
     * @param variable Variable to declare
     * @throws Exception When variable with same name already exists in this context this function throws an exception
     */
    public void declare(Variable variable) throws Exception {
        if (isLocalVariableDeclared(variable.getName())) {
            throw new Exception(String.format("Variable '%s' already exists in this context", variable.getName()));
        }

        variables.put(variable.getName(), variable);
    }

    /**
     * Declares new label in this context
     * @param label Label to declare
     */
    public void declare(Label label) throws Exception {
        if (isLocalLabelDeclared(label.getName())) {
            throw new Exception(String.format("Label '%s' already exists in this context", label.getName()));
        }

        labels.put(label.getName(), label);
    }

    /**
     * Returns whether a variable with the given name is declared locally
     * @param name Variable name to look for
     * @return True, if a variable with the given name is declared locally, otherwise false
     */
    public boolean isLocalTypeDeclared(String name) {
        return types.containsKey(name); 
    }

    /**
     * Returns whether a function with the given name is declared locally
     * @param name Function name to look for
     * @return True, if a function with the given name is declared locally, otherwise false
     */
    public boolean isLocalFunctionDeclared(String name) {
        return functions.containsKey(name);
    }

    /**
     * Returns whether a variable with the given name is declared locally
     * @param name Variable name to look for
     * @return True, if a variable with the given name is declared locally, otherwise false
     */
    public boolean isLocalVariableDeclared(String name) {
        return variables.containsKey(name);
    }

    /**
     * Returns whether a label with the given name is declared locally
     * @param name Label name to look for
     * @return True, if a label with the given name is declared locally, otherwise false
     */
    public boolean isLocalLabelDeclared(String name) {
        return labels.containsKey(name);
    }

    /**
     * Returns whether a variable with the given name is declared locally or globally
     * @param name Variable name to look for
     * @return True, if a variable with the given name is declared locally or globally, otherwise false
     */
    public boolean isVariableDeclared(String name) {
        return variables.containsKey(name) || (context != null && context.isVariableDeclared(name));
    }

    /**
     * Returns whether a type with the given name is declared locally or globally
     * @param name Type name to look for
     * @return True, if a type with the given name is declared locally or globally, otherwise false
     */
    public boolean isTypeDeclared(String name) {
        return types.containsKey(name) || (context != null && context.isTypeDeclared(name));
    }

    /**
     * Returns whether a function with the given name is declared locally or globally
     * @param name Function name to look for
     * @return True, if a function with the given name is declared locally or globally, otherwise false
     */
    public boolean isFunctionDeclared(String name) {
        return functions.containsKey(name) || (context != null && context.isFunctionDeclared(name));
    }

    /**
     * Returns whether a label with the given name is declared locally or globally
     * @param name Label name to look for
     * @return True, if a label with the given name is declared locally or globally, otherwise false
     */
    public boolean isLabelDeclared(String name) {
        return labels.containsKey(name) || (context != null && context.isLabelDeclared(name));
    }

    /**
     * Tries to return type by name locally or globally
     * @param name Type name to look for
     * @return Type corresponding to the given name
     * @throws Exception Throws if the type wasn't found
     */
    public Type getType(String name) {
        if (types.containsKey(name)) {
            return types.get(name);
        }
        else if (context != null) {
            return context.getType(name);
        }
        else {
            return null;
            //throw new Exception(String.format("Couldn't find type '%s'", name));
        }
    }

    /**
     * Tries to return function by name locally or globally
     * @param name Function name to look for
     * @return Function corresponding to the given name
     * @throws Exception Throws if the function wasn't found
     */
    public Functions getFunction(String name) {
        if (functions.containsKey(name)) {
            return functions.get(name);
        }
        else if (context != null) {
            return context.getFunction(name);
        }
        else {
            return null;
            //throw new Exception(String.format("Couldn't find function '%s'", name));
        }
    }

    /**
     * Tries to return variable by name locally or globally
     * @param name Variable name to look for
     * @return Variable corresponding to the given name
     * @throws Exception Throws if the variable wasn't found
     */
    public Variable getVariable(String name) {
        if (variables.containsKey(name)) {
            return variables.get(name);
        }
        else if (context != null) {
            return context.getVariable(name);
        }
        else {
            return null;
            //throw new Exception(String.format("Couldn't find variable '%s'", name));
        }
    }

    /**
     * Tries to return label by name locally or globally
     * @param name Label name to look for
     * @return Label corresponding to the given name
     * @throws Exception Throws if the label wasn't found
     */
    public Label getLabel(String name) {
        if (labels.containsKey(name)) {
            return labels.get(name);
        }
        else if (context != null) {
            return context.getLabel(name);
        }
        else {
            return null;
        }
    }

    /**
     * Returns all variables this context owns
     * @return All variables this context owns
     */
    public Collection<Variable> getVariables() {
        return variables.values();
    }

    /**
     * Returns all functions this context owns
     * @return All functions this context owns
     */
    public Collection<Functions> getFunctions() {
        return functions.values();
    }

    /**
     * Returns all types this context owns
     * @return All types this context owns
     */
    public Collection<Type> getTypes() {
        return types.values();
    }

    /**
     * Returns all labels this context owns
     * @return All labels this context owns
     */
    public Collection<Label> getLabels() {
        return labels.values();
    }

    /**
     * Returns the type context that this context is part of
     * @return Success: Type context that this context is part of, Failure: null
     */
    public Type getTypeParent() {
        if (this instanceof Type) {
            return (Type)this;
        }
        else if (context == null) {
            return null;
        }

        return (context instanceof Type) ? (Type)context : context.getTypeParent();
    }

    /**
     * Returns the function context that this context is part of
     * @return Success: Function context that this context is part of, Failure: null
     */
    public Function getFunctionParent() {
        if (this instanceof Function) {
            return (Function)this;
        }
        else if (context == null) {
            return null;
        }

        return (context instanceof Function) ? (Function)context : context.getFunctionParent();
    }

    /**
     * Returns whether this context is an instance of a type
     * @return True if this context is an instance of a type, otherwise false
     */
    public boolean isType() {
        return (this instanceof Type);
    }

    /**
     * Returns whether this context is an instance of a function
     * @return True if this context is an instance of a function, otherwise false
     */
    public boolean isFunction() {
        return (this instanceof Function);
    }
}

File: Zigzag/Parser/Context.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Context
{
	public string Name { get; protected set; } = string.Empty;
	public string Prefix { get; protected set; } = string.Empty;
	public string Postfix { get; protected set; } = string.Empty;

	public Context? Parent { get; set; }
	public List<Context> Subcontexts { get; private set; } = new List<Context>();

	public bool IsGlobal => GetTypeParent() == null;
	public bool IsMember => !IsGlobal;
	public bool IsType => this is Type;
	public bool IsFunction => this is FunctionImplementation;
	public bool IsInsideFunction => IsFunction || GetFunctionParent() != null;
	public bool IsInsideType => IsType || GetTypeParent() != null;

	public Dictionary<string, Variable> Variables { get; } = new Dictionary<string, Variable>();
	public Dictionary<string, FunctionList> Functions { get; } = new Dictionary<string, FunctionList>();
	public Dictionary<string, Type> Types { get; } = new Dictionary<string, Type>();
	public Dictionary<string, Label> Labels { get; } = new Dictionary<string, Label>();

	/// <summary>
	/// Returns whether the given context is a parent context or higher to this context
	/// </summary>
	public bool IsInside(Context context)
	{
		return Equals(context, this) || (Parent?.IsInside(context) ?? false);
	}

	/// <summary>
	/// Returns the global name of the context
	/// </summary>
	/// <returns>Global name of the context</returns>
	public virtual string GetFullname()
	{
		var parent = Parent?.GetFullname() ?? string.Empty;
		parent = string.IsNullOrEmpty(parent) ? parent : parent + '_';

		var prefix = Prefix.ToLower();
		prefix = string.IsNullOrEmpty(prefix) ? string.Empty : prefix;

		var name = Name.ToLower();
		name = string.IsNullOrEmpty(name) ? name : '_' + name;

		var postfix = Postfix.ToLower();
		postfix = string.IsNullOrEmpty(postfix) ? postfix : '_' + postfix;

		return parent + prefix + name + postfix;
	}

	/// <summary>
	/// Updates types, function and variables when new context is linked
	/// </summary>
	public void Update()
	{
		foreach (var variable in Variables.Values)
		{
			if (variable.IsUnresolved)
			{
				var resolvable = (IResolvable?)variable.Type;

				if (resolvable != null)
				{
					// Try to solve the type
					var type = (TypeNode?)resolvable.Resolve(this);
					variable.Type = type?.Type;
				}
			}
		}

		foreach (var type in Types.Values)
		{
			type.Update();
		}

		foreach (var subcontext in Subcontexts)
		{
			subcontext.Update();
		}
	}
	
	/// <summary>
	/// Links this context with the given context, allowing access to the information of the given context
	/// </summary>
	/// <param name="context">Context to link with</param>
	public void Link(Context context)
	{
		Parent = context;
		Parent.Subcontexts.Add(this);
		Update();
	}

	/// <summary>
	/// Moves all types, functions and variables from the given context to this context, and then destroyes the given context
	/// </summary>
	public void Merge(Context context)
	{
		foreach (var (key, value) in context.Types)
		{
			Types.TryAdd(key, value);
		}

		foreach (var (key, value) in context.Functions)
		{
			Functions.TryAdd(key, value);
		}

		foreach (var (key, value) in context.Variables)
		{
			Variables.TryAdd(key, value);
		}

		foreach (var type in context.Types.Values)
		{
			type.Parent = this;
		}

		foreach (var function in context.Functions.Values)
		{
			foreach (var overload in function.Overloads)
			{
				overload.Parent = this;

				foreach (var implementation in overload.Implementations)
				{
					implementation.Parent = this;
				}
			}
		}

		foreach (var variable in context.Variables.Values)
		{
			variable.Context = this;
		}

		Update();

		context.Destroy();
	}

	/// <summary>
	/// Declares a type into the context
	/// </summary>
	/// <param name="type">Type to declare</param>
	public void Declare(Type type)
	{
		if (IsLocalTypeDeclared(type.Name))
		{
			throw new Exception($"Type '{type.Name}' already exists in this context");
		}

		Types.Add(type.Name, type);
	}

	/// <summary>
	/// Declares a function into the context
	/// </summary>
	/// <param name="function">Function to declare</param>
	public void Declare(Function function)
	{
		FunctionList entry;

		if (IsLocalFunctionDeclared(function.Name))
		{
			entry = Functions[function.Name];
		}
		else
		{
			Functions.Add(function.Name, (entry = new FunctionList()));
		}

		entry.Add(function);
	}

	/// <summary>
	/// Declares a variable into the context
	/// </summary>
	/// <param name="variable">Variable to declare</param>
	public void Declare(Variable variable)
	{
		if (IsLocalVariableDeclared(variable.Name))
		{
			throw new Exception($"Variable '{variable.Name}' already exists in this context");
		}

		// Update variable context
		variable.Context = this;

		// Add variable to the list
		Variables.Add(variable.Name, variable);
	}

	/// <summary>
	/// Declares a variable into the context
	/// </summary>
	public void Declare(Type? type, VariableCategory category, string name)
	{
		if (IsLocalVariableDeclared(name))
		{
			throw new Exception($"Variable '{name}' already exists in this context");
		}

		// When a variable is created this way it is automatically declared into this context
		Variable.Create(this, type, category, name, AccessModifier.PUBLIC);
	}

	/// <summary>
	/// Declares a label into the context
	/// </summary>
	/// <param name="label">Label to declare</param>
	public void Declare(Label label)
	{
		if (IsLocalLabelDeclared(label.GetName()))
		{
			throw new Exception($"Label '{label.GetName()}' already exists in this context");
		}

		Labels.Add(label.GetName(), label);
	}

	/// <summary>
	/// Declares an already existing type with different name
	/// </summary>
	public void DeclareTypeAlias(string alias, Type type)
	{
		if (IsLocalTypeDeclared(alias))
		{
			throw new Exception($"Tried to declare type alias '{alias}' but the name was already reserved");
		}

		Types.Add(alias, type);
	}

	public virtual bool IsLocalTypeDeclared(string name)
	{
		return Types.ContainsKey(name);
	}

	public virtual bool IsLocalFunctionDeclared(string name)
	{
		return Functions.ContainsKey(name);
	}

	public virtual bool IsLocalVariableDeclared(string name)
	{
		return Variables.ContainsKey(name);
	}

	public virtual bool IsLocalLabelDeclared(string name)
	{
		return Labels.ContainsKey(name);
	}

	public virtual bool IsVariableDeclared(string name)
	{
		return Variables.ContainsKey(name) || (Parent != null && Parent.IsVariableDeclared(name));
	}

	public virtual bool IsTypeDeclared(string name)
	{
		return Types.ContainsKey(name) || (Parent != null && Parent.IsTypeDeclared(name));
	}

	public virtual bool IsTemplateTypeDeclared(string name)
	{
		return IsTypeDeclared(name) && GetType(name)!.IsTemplateType;
	}

	public virtual bool IsFunctionDeclared(string name)
	{
		return Functions.ContainsKey(name) || (Parent != null && Parent.IsFunctionDeclared(name));
	}

	public virtual bool IsLabelDeclared(string name)
	{
		return Labels.ContainsKey(name) || (Parent != null && Parent.IsLabelDeclared(name));
	}

	public Type? GetType(string name)
	{
		if (Types.ContainsKey(name))
		{
			return Types[name];
		}
		else if (Parent != null)
		{
			return Parent.GetType(name);
		}
		else
		{
			return null;
		}
	}

	public virtual FunctionList? GetFunction(string name)
	{
		if (Functions.ContainsKey(name))
		{
			return Functions[name];
		}
		else if (Parent != null)
		{
			return Parent.GetFunction(name);
		}
		else
		{
			return null;
		}
	}

	public virtual Variable? GetVariable(string name)
	{
		if (Variables.ContainsKey(name))
		{
			return Variables[name];
		}
		else if (Parent != null)
		{
			return Parent.GetVariable(name);
		}
		else
		{
			return null;
		}
	}

	public virtual Label? GetLabel(string name)
	{
		if (Labels.ContainsKey(name))
		{
			return Labels[name];
		}
		else if (Parent != null)
		{
			return Parent.GetLabel(name);
		}
		else
		{
			return null;
		}
	}

	public Type? GetTypeParent()
	{
		if (IsType)
		{
			return (Type)this;
		}

		return Parent?.GetTypeParent();
	}

	public FunctionImplementation? GetFunctionParent()
	{
		if (IsFunction)
		{
			return (FunctionImplementation)this;
		}

		return Parent?.GetFunctionParent();
	}

	public virtual IEnumerable<FunctionImplementation> GetImplementedFunctions()
	{
		return Functions.Values
			.SelectMany(f => f.Overloads)
			.SelectMany(f => f.Implementations)
			.Where(i => i.Node != null);
	}
	
	public virtual IEnumerable<FunctionImplementation> GetFunctionImplementations()
	{
		return Functions.Values
			.SelectMany(f => f.Overloads)
			.SelectMany(f => f.Implementations);
	}

	public void Destroy()
	{
		Parent?.Subcontexts.Remove(this);
		Parent = null;
	}

	public override bool Equals(object? obj)
	{
		return obj is Context context &&
			   Name == context.Name &&
			   EqualityComparer<List<Context>>.Default.Equals(Subcontexts, context.Subcontexts) &&
			   IsType == context.IsType &&
			   IsFunction == context.IsFunction &&
			   EqualityComparer<Dictionary<string, Variable>>.Default.Equals(Variables, context.Variables) &&
			   EqualityComparer<Dictionary<string, FunctionList>>.Default.Equals(Functions, context.Functions) &&
			   EqualityComparer<Dictionary<string, Type>>.Default.Equals(Types, context.Types) &&
			   EqualityComparer<Dictionary<string, Label>>.Default.Equals(Labels, context.Labels);
	}

	public override int GetHashCode()
	{
		HashCode hash = new HashCode();
		hash.Add(Name);
		hash.Add(Subcontexts);
		hash.Add(IsType);
		hash.Add(IsFunction);
		hash.Add(Variables);
		hash.Add(Functions);
		hash.Add(Types);
		hash.Add(Labels);
		return hash.ToHashCode();
	}
}